File: packages/akribes-sdk-python/akribes_sdk/resources/projects.py
from __future__ import annotations

from urllib.parse import quote

from akribes_sdk._pagination import AsyncPage
from akribes_sdk._parsers import parse_project, parse_script
from akribes_sdk.errors import NotFoundError
from akribes_sdk.models import Project, Script, ScriptChannel
from akribes_sdk.resources._base import Resource
from akribes_sdk.resources._sentinel import _MISSING


class Projects(Resource):

    def list(self) -> AsyncPage[Project]:
        async def fetch(offset: int, limit: int) -> tuple[list[Project], bool]:
            res = await self._request(
                "GET",
                f"{self._base_url}/projects",
                params={"limit": limit, "offset": offset},
            )
            items = [parse_project(p) for p in res.json()]
            return items, len(items) == limit
        return AsyncPage(fetch)

    async def get(self, project_id: int, *, default=_MISSING) -> Project:
        """Return the project or raise :class:`NotFoundError`.

        Pass ``default=None`` (or any other value) to suppress the error and
        return that value instead — mirroring ``dict.get(key, default)``.
        """
        try:
            res = await self._request("GET", f"{self._base_url}/projects/{project_id}")
        except NotFoundError:
            if default is _MISSING:
                raise
            return default  # type: ignore[return-value]
        return parse_project(res.json())

    async def create(self, name: str) -> Project:
        res = await self._request("POST", f"{self._base_url}/projects", json={"name": name})
        return parse_project(res.json())

    async def update(self, project_id: int, name: str) -> Project:
        res = await self._request("PATCH", f"{self._base_url}/projects/{project_id}", json={"name": name})
        return parse_project(res.json())

    async def delete(self, project_id: int) -> None:
        await self._request("DELETE", f"{self._base_url}/projects/{project_id}")

    async def duplicate(self, project_id: int) -> Project:
        """Duplicate a project (including all scripts). Server picks the copy
        name. Requires a wildcard-scoped identity."""
        res = await self._request(
            "POST", f"{self._base_url}/projects/{project_id}/duplicate", json={}
        )
        return parse_project(res.json())

    async def reorder(self, order: list[int]) -> None:
        """Set the global project ordering. ``order`` is the list of project
        IDs in the desired order. Requires a wildcard-scoped identity."""
        await self._request(
            "PUT", f"{self._base_url}/projects/reorder", json={"order": order}
        )

    # ── Flat cross-project script ops ───────────────────────────────────
    #
    # The project-scoped chain (``client.scripts.X``) is the primary surface
    # for script management, but admin-style code that touches several
    # projects in a row reads more naturally with flat, cross-project ops on
    # ``projects`` itself — same as Rust ``projects().X_script(...)`` and
    # TypeScript ``projects.XScript(...)``.

    def _script_url(self, project_id: int, script_name: str, *segments: str) -> str:
        encoded = quote(script_name, safe="")
        url = f"{self._base_url}/projects/{project_id}/scripts/{encoded}"
        for s in segments:
            url = f"{url}/{s}"
        return url

    async def list_scripts(self, project_id: int) -> list[Script]:
        """List scripts in a specific project. Flat alternative to
        ``client.scripts.list()`` (which uses the client's bound project)."""
        res = await self._request(
            "GET", f"{self._base_url}/projects/{project_id}/scripts"
        )
        return [parse_script(s) for s in res.json()]

    async def move_script(
        self,
        project_id: int,
        script_name: str,
        target_project_id: int,
    ) -> Script:
        """Move a script from ``project_id`` to ``target_project_id``."""
        res = await self._request(
            "POST",
            self._script_url(project_id, script_name, "move"),
            json={"target_project_id": target_project_id},
        )
        return parse_script(res.json())

    async def rename_script(
        self,
        project_id: int,
        current_name: str,
        new_name: str,
    ) -> None:
        """Rename a script in ``project_id``."""
        await self._request(
            "PATCH",
            self._script_url(project_id, current_name),
            json={"new_name": new_name},
        )

    async def delete_script(self, project_id: int, script_name: str) -> None:
        """Delete a script in ``project_id``."""
        await self._request("DELETE", self._script_url(project_id, script_name))

    async def duplicate_script(
        self,
        project_id: int,
        script_name: str,
    ) -> Script:
        """Duplicate a script within ``project_id``. The server picks the
        copy name."""
        res = await self._request(
            "POST",
            self._script_url(project_id, script_name, "duplicate"),
            json={},
        )
        return parse_script(res.json())

    async def list_channels(
        self,
        project_id: int,
        script_name: str,
    ) -> list[ScriptChannel]:
        """List channels for ``script_name`` in ``project_id`` (#1141).

        Flat cross-project alternative to
        ``client.channels.list(script_name)`` — mirrors Rust
        ``projects.list_channels`` and TS ``projects.listChannels``."""
        from akribes_sdk._parsers import parse_script_channel
        res = await self._request(
            "GET",
            self._script_url(project_id, script_name, "channels"),
        )
        return [parse_script_channel(c) for c in res.json()]

    async def reorder_scripts(
        self,
        project_id: int,
        order: list[int],
    ) -> None:
        """Reorder scripts within ``project_id`` (#1141).

        ``order`` is the list of script IDs in the desired order. Flat
        cross-project alternative to ``client.scripts.reorder(order)``."""
        await self._request(
            "PUT",
            f"{self._base_url}/projects/{project_id}/scripts/reorder",
            json={"order": order},
        )

File: crates/akribes-sdk/src/sub/projects.rs
use std::sync::Arc;

use crate::client::{AkribesClient, Inner};
use crate::error::Result;
use crate::models::*;

/// Sub-client for project management. Obtained via [`AkribesClient::projects()`].
///
/// Does **not** require a `project_id` on the parent client.
#[derive(Clone, Debug)]
pub struct ProjectsClient {
    pub(crate) inner: Arc<Inner>,
}

impl ProjectsClient {
    pub(crate) fn new(inner: Arc<Inner>) -> Self {
        Self { inner }
    }

    /// Wrap as an AkribesClient to reuse HTTP helpers.
    fn c(&self) -> AkribesClient {
        AkribesClient {
            inner: Arc::clone(&self.inner),
        }
    }

    pub async fn list(&self) -> Result<Vec<Project>> {
        let url = format!("{}/projects", self.inner.base_url);
        self.c().get_list(&url).await
    }

    pub async fn get(&self, project_id: i64) -> Result<Option<Project>> {
        let url = format!("{}/projects/{}", self.inner.base_url, project_id);
        self.c().get_opt(&url).await
    }

    pub async fn create(&self, name: &str) -> Result<Project> {
        let url = format!("{}/projects", self.inner.base_url);
        self.c().post(&url, &CreateProjectRequest { name }).await
    }

    pub async fn update(&self, project_id: i64, name: &str) -> Result<Project> {
        let url = format!("{}/projects/{}", self.inner.base_url, project_id);
        self.c().patch(&url, &UpdateProjectRequest { name }).await
    }

    /// Look up a project by numeric id (if `id_or_name` parses as `i64`) or
    /// by exact name. Returns `None` when no project matches.
    ///
    /// Numeric-looking inputs are treated as ids first; a caller that really
    /// wants to look up a project *named* e.g. `"42"` must currently fetch
    /// the full list and filter themselves.
    pub async fn resolve(&self, id_or_name: &str) -> Result<Option<Project>> {
        if let Ok(id) = id_or_name.parse::<i64>() {
            return self.get(id).await;
        }
        let list = self.list().await?;
        Ok(list.into_iter().find(|p| p.name == id_or_name))
    }

    pub async fn delete(&self, project_id: i64) -> Result<()> {
        let url = format!("{}/projects/{}", self.inner.base_url, project_id);
        self.c().delete(&url).await?;
        Ok(())
    }

    /// Duplicate a project (including all scripts). Server picks the copy
    /// name. Requires a wildcard-scoped identity.
    pub async fn duplicate(&self, project_id: i64) -> Result<Project> {
        let url = format!("{}/projects/{}/duplicate", self.inner.base_url, project_id);
        self.c().post(&url, &serde_json::json!({})).await
    }

    /// Set the global project ordering. `order` is the list of project IDs
    /// in the desired order. Requires a wildcard-scoped identity.
    pub async fn reorder(&self, order: Vec<i64>) -> Result<()> {
        let url = format!("{}/projects/reorder", self.inner.base_url);
        self.c().put_empty(&url, &ReorderRequest { order }).await
    }

    // ── Flat cross-project script ops ───────────────────────────────────
    //
    // The project-scoped chain (`client.project(id).scripts().X`) is the
    // primary surface for script management, but admin-style code that
    // touches several projects in a row reads more naturally with flat,
    // cross-project ops on `projects` itself — same as TS `projects.*`.
    // Both surfaces coexist; these delegate to the equivalent server
    // endpoints directly to avoid an extra constructor allocation.

    fn script_url(&self, project_id: i64, script_name: &str) -> String {
        format!(
            "{}/projects/{}/scripts/{}",
            self.inner.base_url,
            project_id,
            urlencoding::encode(script_name)
        )
    }

    /// List scripts in a specific project. Flat alternative to
    /// `client.project(id).scripts().list()`.
    pub async fn list_scripts(&self, project_id: i64) -> Result<Vec<Script>> {
        let url = format!("{}/projects/{}/scripts", self.inner.base_url, project_id);
        self.c().get_list(&url).await
    }

    /// Move a script from `src_project_id` to `dest_project_id`. Flat
    /// alternative to `client.project(src).scripts().move_to(name, dest)`.
    pub async fn move_script(
        &self,
        src_project_id: i64,
        src_script_name: &str,
        dest_project_id: i64,
    ) -> Result<Script> {
        let url = format!("{}/move", self.script_url(src_project_id, src_script_name));
        self.c()
            .post(
                &url,
                &MoveScriptRequest {
                    target_project_id: dest_project_id,
                },
            )
            .await
    }

    /// Rename a script in `project_id`. Flat alternative to
    /// `client.project(id).scripts().rename(current, new)`.
    pub async fn rename_script(
        &self,
        project_id: i64,
        current_name: &str,
        new_name: &str,
    ) -> Result<()> {
        let url = self.script_url(project_id, current_name);
        self.c()
            .patch_empty(&url, &RenameScriptRequest { new_name })
            .await
    }

    /// Delete a script in `project_id`. Flat alternative to
    /// `client.project(id).scripts().delete(name)`.
    pub async fn delete_script(&self, project_id: i64, script_name: &str) -> Result<()> {
        let url = self.script_url(project_id, script_name);
        self.c().delete(&url).await?;
        Ok(())
    }

    /// Duplicate a script within `project_id`. The server picks the copy
    /// name; `new_name` is accepted for parity with the TypeScript SDK and
    /// will be honored once the server supports it. Flat alternative to
    /// `client.project(id).scripts().duplicate(name)`.
    pub async fn duplicate_script(
        &self,
        project_id: i64,
        script_name: &str,
        _new_name: Option<&str>,
    ) -> Result<Script> {
        let url = format!("{}/duplicate", self.script_url(project_id, script_name));
        self.c().post(&url, &serde_json::json!({})).await
    }

    /// List channels for `script_name` in `project_id` (#1141). Flat
    /// cross-project alternative to
    /// `client.project(id).channels().list(name)`. Mirrors TS
    /// `projects.listChannels`.
    pub async fn list_channels(
        &self,
        project_id: i64,
        script_name: &str,
    ) -> Result<Vec<ScriptChannel>> {
        let url = format!("{}/channels", self.script_url(project_id, script_name));
        self.c().get_list(&url).await
    }

    /// Reorder scripts within `project_id` (#1141). `order` is the list of
    /// script IDs in the desired order. Flat cross-project alternative to
    /// `client.project(id).scripts().reorder(order)`. Mirrors TS
    /// `projects.reorderScripts`.
    pub async fn reorder_scripts(&self, project_id: i64, order: Vec<i64>) -> Result<()> {
        let url = format!(
            "{}/projects/{}/scripts/reorder",
            self.inner.base_url, project_id
        );
        self.c().put_empty(&url, &ReorderRequest { order }).await
    }
}
